### PR TITLE
Fixed the bug in GPU tests

### PR DESCRIPTION
**Pull request by juhlee-microsoft**
- This changes are only in Ubuntu 16.04 with supported VMSize. 
- Removed Copy-and-Paste the same test script execution with different VM size. But use OverriderVMSize parameter.

This is DRAFT PR not actual. @chvalean will merge this change to his work. No plan to merge this PR.

Open items: https://docs.microsoft.com/en-us/azure/virtual-machines/linux/n-series-driver-setup?toc=%2fazure%2fvirtual-machines%2flinux%2ftoc.json

### Test Results

<Unsupported VMSize>
.\Run-LisaV2.ps1 -XMLSecretFile 'C:\Users\juhlee\source\repo\AzureSecrets.xml' -TestLocation 'westus2' -RGIdentifier 'juhlee-gpu' -TestPlatform 'Azure' -StorageAccount 'ExistingStorage_Standard' -ARMImageName 'Canonical UbuntuServer 16.04-LTS latest' -TestNames 'VERIFY-NVIDIA-CUDA-DRIVER' -OverrideVMSize "Standard_DS1_v2"

[LISAv2 Test Results Summary]
Test Run On           : 02/16/2019 21:03:17
ARM Image Under Test  : Canonical : UbuntuServer : 16.04-LTS : latest
Total Test Cases      : 1 (0 Passed, 0 Failed, 1 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:4

   ID TestCaseName                                                                TestResult TestDuration(in minutes)
---------------------------------------------------------------------------------------------------------------------
    1 VERIFY-NVIDIA-CUDA-DRIVER                                                      ABORTED                 2.31

Logs can be found at C:\LISAv2\UL28\TestResults\2019-16-02-13-03-14-4613

02/16/2019 21:07:42 : [INFO ] Current Git Branch:  juhlee/simple_gpu_tc
02/16/2019 21:07:42 : [INFO ] Current Git Commit ID: 71ee1c9cc9280f6482d64e72db6b6c76d32104c1
02/16/2019 21:07:42 : [INFO ] Creating 'C:\LISAv2\UL28\Azure-UL28-TestLogs.zip' from 'C:\LISAv2\UL28\TestResults\2019-16-02-13-03-14-4613'
02/16/2019 21:07:42 : [INFO ] C:\LISAv2\UL28\Azure-UL28-TestLogs.zip created successfully.
02/16/2019 21:07:42 : [INFO ] Analyzing results..
02/16/2019 21:07:42 : [INFO ] Copying all files back to original working directory: C:\Users\juhlee\source\repo\LISAv2.
02/16/2019 21:07:46 : [INFO ] Setting workspace back to original location: C:\Users\juhlee\source\repo\LISAv2
02/16/2019 21:07:46 : [INFO ] LISAv2 exit code: 1

-->Capture the EXCEPTION log below.
02/16/2019 21:05:39 : [ERROR] EXCEPTION : Found unsupported VMSize Standard_DS1_v2. It shall be in Standard_NC6 Standard_NC6s_v2 Standard_NV6s_v3 Standard_NV6 Standard_NV6s_v2 Standard_NC12 Standard_NC12s_v2
Standard_NC12s_v3 Standard_NV12 Standard_NV12s_v2 Standard_NC24 Standard_NC24s_v2 Standard_NC24s_v3 Standard_NV24 Standard_NV24s_v2 Standard_NC24r Standard_NC24rs_v2 Standard_NC24rs_v3 at line: 21
02/16/2019 21:05:39 : [ERROR] EXCEPTION FULL: System.Management.Automation.RuntimeException: Found unsupported VMSize Standard_DS1_v2. It shall be in Standard_NC6 Standard_NC6s_v2 Standard_NV6s_v3 Standard_NV
6 Standard_NV6s_v2 Standard_NC12 Standard_NC12s_v2 Standard_NC12s_v3 Standard_NV12 Standard_NV12s_v2 Standard_NC24 Standard_NC24s_v2 Standard_NC24s_v3 Standard_NV24 Standard_NV24s_v2 Standard_NC24r Standard_N
C24rs_v2 Standard_NC24rs_v3
02/16/2019 21:05:39 : [INFO ] Test case script result does not match known ones: Aborted
02/16/2019 21:05:39 : [ERROR] Failed to retrieve a known test result: Aborted

<Supported VMSize>

.\Run-LisaV2.ps1 -XMLSecretFile 'C:\Users\juhlee\source\repo\AzureSecrets.xml' -TestLocation 'westus2' -RGIdentifier 'juhlee-gpu' -TestPlatform 'Azure' -StorageAccount 'ExistingStorage_Standard' -ARMImageName 'Canonical UbuntuServer 16.04-LTS latest' -TestNames 'VERIFY-NVIDIA-CUDA-DRIVER' -OverrideVMSize "Standard_NC24r"

[LISAv2 Test Results Summary]
Test Run On           : 02/16/2019 21:12:55
ARM Image Under Test  : Canonical : UbuntuServer : 16.04-LTS : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:15

   ID TestCaseName                                                                TestResult TestDuration(in minutes)
---------------------------------------------------------------------------------------------------------------------
    1 VERIFY-NVIDIA-CUDA-DRIVER                                                         PASS                13.46
        PCI Express pass-through : PASS
        lspci : PASS
        lshw -c video : PASS
        nvidia-smi : PASS

Logs can be found at C:\LISAv2\JA93\TestResults\2019-16-02-13-12-52-0902

02/16/2019 21:28:34 : [INFO ] Current Git Branch:  juhlee/simple_gpu_tc
02/16/2019 21:28:34 : [INFO ] Current Git Commit ID: 71ee1c9cc9280f6482d64e72db6b6c76d32104c1
02/16/2019 21:28:34 : [INFO ] Creating 'C:\LISAv2\JA93\Azure-JA93-TestLogs.zip' from 'C:\LISAv2\JA93\TestResults\2019-16-02-13-12-52-0902'
02/16/2019 21:28:34 : [INFO ] C:\LISAv2\JA93\Azure-JA93-TestLogs.zip created successfully.
02/16/2019 21:28:34 : [INFO ] Analyzing results..
02/16/2019 21:28:34 : [INFO ] Copying all files back to original working directory: C:\Users\juhlee\source\repo\LISAv2.
02/16/2019 21:28:38 : [INFO ] Cleaning up C:\LISAv2\JA93
02/16/2019 21:28:38 : [INFO ] Setting workspace back to original location: C:\Users\juhlee\source\repo\LISAv2
02/16/2019 21:28:38 : [INFO ] LISAv2 exit code: 0

--> No EXCEPTION but success log.
02/16/2019 21:15:18 : [INFO ] Found supported VMSize successfully: Standard_NC24r

### CODE REVIEW

File: Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1 [hunk 1]
```diff
@@ -9,8 +9,18 @@ function Main {
     # Create test result
     $resultArr = @()
     $currentTestResult = Create-TestResultObject
+    $superUser = "root"
     try {
         $clientVMData = $allVMData
+        $supportVMSizes = @("Standard_NC6","Standard_NC6s_v2","Standard_NV6s_v3","Standard_NV6","Standard_NV6s_v2","Standard_NC12","Standard_NC12s_v2","Standard_NC12s_v3","Standard_NV12","Standard_NV12s_v2","Standard_NC24"  ,"Standard_NC24s_v2","Standard_NC24s_v3","Standard_NV24","Standard_NV24s_v2","Standard_NC24r","Standard_NC24rs_v2","Standard_NC24rs_v3")
```
Comment (iamshital) [inline Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1]: Split this line please.

File: Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1 [hunk 2]
```diff
@@ -55,30 +65,37 @@ collect_VM_properties
             if ($TestProvider.RestartAllDeployments($clientVMData)) {
                 if (($clientVMData.InstanceSize.Contains("Standard_NC6")) -or
                         ($clientVMData.InstanceSize.Contains("Standard_NC6s_v2")) -or
-                        ($clientVMData.InstanceSize.Contains("Standard_NV6"))) {
+                        ($clientVMData.InstanceSize.Contains("Standard_NV6s_v3")) -or
+                        ($clientVMData.InstanceSize.Contains("Standard_NV6")) -or
+                        ($clientVMData.InstanceSize.Contains("Standard_NV6s_v2"))) {
```
Comment (iamshital) [inline Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1]: Going forward, using regex is more helpful...

File: Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1 [hunk 3]
```diff
@@ -55,30 +65,37 @@ collect_VM_properties
...
+                        ($clientVMData.InstanceSize.Contains("Standard_NV6")) -or
+                        ($clientVMData.InstanceSize.Contains("Standard_NV6s_v2"))) {
                     $expectedCount = 1
                 }
                 elseif (($clientVMData.InstanceSize.Contains("Standard_NC12")) -or
                         ($clientVMData.InstanceSize.Contains("Standard_NC12s_v2")) -or
-                        ($clientVMData.InstanceSize.Contains("Standard_NV12"))) {
+                        ($clientVMData.InstanceSize.Contains("Standard_NC12s_v3")) -or
+                        ($clientVMData.InstanceSize.Contains("Standard_NV12")) -or
+                        ($clientVMData.InstanceSize.Contains("Standard_NV12s_v2"))) {
```
Comment (iamshital) [inline Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1]: Going forward, using regex is more helpful...

File: Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1 [hunk 4]
```diff
@@ -55,30 +65,37 @@ collect_VM_properties
...
-                    $expectedCount = 4
-                }
-                elseif (($clientVMData.InstanceSize.Contains("Standard_NC24r")) -or
-                        ($clientVMData.InstanceSize.Contains("Standard_NC24rs_v2"))) {
+                        ($clientVMData.InstanceSize.Contains("Standard_NC24s_v3")) -or
+                        ($clientVMData.InstanceSize.Contains("Standard_NV24")) -or
+                        ($clientVMData.InstanceSize.Contains("Standard_NV24s_v2")) -or
+                        ($clientVMData.InstanceSize.Contains("Standard_NC24r")) -or
+                        ($clientVMData.InstanceSize.Contains("Standard_NC24rs_v2")) -or
+                        ($clientVMData.InstanceSize.Contains("Standard_NC24rs_v3"))) {
```
Comment (iamshital) [inline Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1]: Going forward, using regex is more helpful...

File: Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1 [hunk 5]
```diff
@@ -92,10 +109,11 @@ collect_VM_properties
                 #endregion
 
                 #region lspci
-                $lspci = Run-LinuxCmd -ip $clientVMData.PublicIP -port $clientVMData.SSHPort -username "root" -password $password -command "lspci" -ignoreLinuxExitCode
+                $lspci = Run-LinuxCmd -ip $clientVMData.PublicIP -port $clientVMData.SSHPort -username $superUser -password $password -command "lspci" -ignoreLinuxExitCode
                 Set-Content -Value $lspci -Path $LogDir\lspci.txt -Force
-                if ((Select-String -Path $LogDir\lspci.txt -Pattern "NVIDIA Corporation").Matches.Count -eq $expectedCount) {
-                    Write-LogInfo "Expected `"3D controller: NVIDIA Corporation`" count: $expectedCount. Observed Count: $expectedCount"
+                $actualCount = (Select-String -Path $LogDir\lspci.txt -Pattern "NVIDIA Corporation").Matches.Count
+                Write-LogInfo "Expected `"3D controller: NVIDIA Corporation`" count: 1. Observed Count: $actualCount"
```
Comment (iamshital) [inline Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1]: Wait, Expected count is fixed now?

**Comment by chvalean** [comment]
@juhlee-microsoft  no need to review this, these will get in the other PR

File: Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1 (hunk 1 above)
Comment (chvalean) [inline Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1]: @juhlee-microsoft I don't like to hardcode VM sizes in the scripts.
Can we go with the same approach as for nvme, where the number of disks is computed from the number of cores divided by 8?
For GPU, these would be a 6 to 1 ratio (1 GPU for each 6 cores).

Comment (juhlee-microsoft) [reply Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1]: Will work on regex style

File: Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1 (hunk 5 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1]: Even closed, I’d like to comment back. I found that number is fixed, 1 regardless VM sizes.

File: Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1 (hunk 1 above)
Comment (chvalean) [reply Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1]: > Will work on regex style

with the above plan to compute the GPUs based on # of cores, we don't need any regex.
Will implement the above in my PR.

**Comment by juhlee-microsoft** [comment]
Ok. Let me close this PR then. Thanks.

**Pull request closed without merging**